File: state_machine.py
"""
state_machine.py
================
Sprint 2 — Temporal Smoothing & Stable Advice State Machine.

Internal states (5):    STOP | UNCERTAIN | WAIT | SAFE | WALK_FAST
Spoken outputs (4):     "Stop" | "Please wait" | "Cross now" | "Walk faster"

State transitions are gated by an N-frame hysteresis buffer:
  • A new state must persist for `smoothing_frames` consecutive frames
    before it is committed.
  • Low-confidence frames immediately force UNCERTAIN (safety-first).
  • Secondary direction cues appended only when confidence is high.
"""

from dataclasses import dataclass, field

# ─────────────────────────────────────────────────────────────────────────────
# State constants
# ─────────────────────────────────────────────────────────────────────────────
STATE_STOP      = "STOP"
STATE_UNCERTAIN = "UNCERTAIN"
STATE_WAIT      = "WAIT"
STATE_SAFE      = "SAFE"
STATE_WALK_FAST = "WALK_FAST"

ALL_STATES = [STATE_STOP, STATE_UNCERTAIN, STATE_WAIT, STATE_SAFE, STATE_WALK_FAST]

# spoken text + audio file stem for each internal state
_SPOKEN_MAP: dict[str, tuple[str, str]] = {
    STATE_STOP:      ("Stop",        "stop"),
    STATE_UNCERTAIN: ("Please wait", "stop"),
    STATE_WAIT:      ("Please wait", "stop"),
    STATE_SAFE:      ("Cross now",   "walk_normal"),
    STATE_WALK_FAST: ("Walk faster", "walk_fast"),
}


# ─────────────────────────────────────────────────────────────────────────────
# Output struct
# ─────────────────────────────────────────────────────────────────────────────
@dataclass
class StateOutput:
    internal_state: str          # one of ALL_STATES
    spoken_text:    str          # "Stop" / "Please wait" / "Cross now" / "Walk faster"
    audio_stem:     str          # filename stem inside voice_cache/
    confidence:     float        # 0.0–1.0
    risk_score:     float = 0.0  # raw risk value from risk scorer
    secondary_cue:  str   = ""   # optional: "Raise your right hand." etc.

    @property
    def full_spoken(self) -> str:
        """Full spoken sentence including any secondary cue."""
        if self.secondary_cue:
            return f"{self.spoken_text}. {self.secondary_cue}"
        return self.spoken_text


# ─────────────────────────────────────────────────────────────────────────────
# State machine
# ─────────────────────────────────────────────────────────────────────────────
class SafetyStateMachine:
    """
    Hysteresis-gated state machine for crossing safety advice.

    A proposed state must persist for `smoothing_frames` consecutive frames
    before it is committed to the output. This prevents rapid flickering.

    UNCERTAIN is committed immediately on any low-confidence frame —
    it is always safer to pause than to give wrong guidance.
    """

    def __init__(self, smoothing_frames: int = 8, min_confidence: float = 0.70):
        self.smoothing_frames   = smoothing_frames
        self.min_confidence     = min_confidence

        self._current_state = STATE_SAFE
        self._pending_state = STATE_SAFE
        self._pending_count = 0

    # ── public API ────────────────────────────────────────────────────
    @property
    def current_state(self) -> str:
        return self._current_state

    def update(
        self,
        proposed_state: str,
        confidence:     float,
        risk_score:     float = 0.0,
        secondary_cue:  str   = "",
    ) -> StateOutput:
        """
        Apply hysteresis gate and return a StateOutput.

        Args:
            proposed_state: Raw state from risk scorer.
            confidence:     0.0–1.0 from risk scorer.
            risk_score:     0.0–1.0 raw risk value (for logging / display).
            secondary_cue:  Direction hint string (empty if confidence low).

        Returns:
            StateOutput with the gated committed state.
        """
        assert proposed_state in ALL_STATES, f"Unknown state: {proposed_state}"

        # Low confidence → UNCERTAIN immediately (safety-first, no gate)
        if confidence < self.min_confidence:
            self._current_state = STATE_UNCERTAIN
            self._pending_state = STATE_UNCERTAIN
            self._pending_count = 0
            spoken, stem = _SPOKEN_MAP[STATE_UNCERTAIN]
            return StateOutput(
                internal_state = STATE_UNCERTAIN,
                spoken_text    = spoken,
                audio_stem     = stem,
                confidence     = confidence,
                risk_score     = risk_score,
                secondary_cue  = "",
            )

        # Same as current — reset pending accumulator
        if proposed_state == self._current_state:
            self._pending_state = proposed_state
            self._pending_count = 0

        # Continuing to accumulate the same pending state
        elif proposed_state == self._pending_state:
            self._pending_count += 1
            if self._pending_count >= self.smoothing_frames:
                self._current_state = proposed_state
                self._pending_count = 0

        # New candidate state — restart accumulation
        else:
            self._pending_state = proposed_state
            self._pending_count = 1

        spoken, stem = _SPOKEN_MAP[self._current_state]

        # Secondary cue only when confident
        cue = secondary_cue if confidence >= self.min_confidence else ""

        return StateOutput(
            internal_state = self._current_state,
            spoken_text    = spoken,
            audio_stem     = stem,
            confidence     = confidence,
            risk_score     = risk_score,
            secondary_cue  = cue,
        )

    def reset(self):
        """Call between clips or camera restarts."""
        self._current_state = STATE_SAFE
        self._pending_state = STATE_SAFE
        self._pending_count = 0

File: pedestrian_safety/lib/domain/usecases/state_machine.dart
import '../../data/models/state_output.dart';
import '../entities/safety_state.dart';

class SafetyStateMachine {
  final int smoothingFrames;
  final double minConfidence;

  SafetyState _currentState = SafetyState.uncertain;
  SafetyState _pendingState = SafetyState.uncertain;
  int _pendingCount = 0;

  SafetyStateMachine({
    this.smoothingFrames = 3,
    this.minConfidence = 0.65,
  });

  SafetyState get currentState => _currentState;

  static final Map<SafetyState, (String, String)> _spokenMap = {
    SafetyState.stop: ('Stop', 'stop'),
    SafetyState.uncertain: ('Please wait', 'stop'),
    SafetyState.wait: ('Please wait', 'stop'),
    SafetyState.safe: ('Cross now', 'walk_normal'),
    SafetyState.walkFast: ('Walk faster', 'walk_fast'),
  };

  int _threatLevel(SafetyState state) {
    switch (state) {
      case SafetyState.safe:
        return 0;
      case SafetyState.walkFast:
        return 1;
      case SafetyState.wait:
        return 2;
      case SafetyState.uncertain:
        return 3;
      case SafetyState.stop:
        return 4;
    }
  }

  StateOutput update({
    required SafetyState proposedState,
    required double confidence,
    double riskScore = 0.0,
    String secondaryCue = '',
  }) {
    // Low confidence -> UNCERTAIN immediately (safety-first, no gate)
    if (confidence < minConfidence) {
      _currentState = SafetyState.uncertain;
      _pendingState = SafetyState.uncertain;
      _pendingCount = 0;
      final mapData = _spokenMap[SafetyState.uncertain]!;
      return StateOutput(
        internalState: SafetyState.uncertain,
        spokenText: mapData.$1,
        audioStem: mapData.$2,
        confidence: confidence,
        riskScore: riskScore,
        secondaryCue: '',
      );
    }

    // Immediate threat escalation (safety-first, bypass smoothing)
    if (_threatLevel(proposedState) > _threatLevel(_currentState)) {
      _currentState = proposedState;
      _pendingState = proposedState;
      _pendingCount = 0;
    }
    // Same as current — reset pending accumulator
    else if (proposedState == _currentState) {
      _pendingState = proposedState;
      _pendingCount = 0;
    }
    // Continuing to accumulate the same pending state (de-escalation smoothing)
    else if (proposedState == _pendingState) {
      _pendingCount++;
      if (_pendingCount >= smoothingFrames) {
        _currentState = proposedState;
        _pendingCount = 0;
      }
    }
    // New candidate state — restart accumulation
    else {
      _pendingState = proposedState;
      _pendingCount = 1;
    }

    final mapData = _spokenMap[_currentState]!;
    final cue = confidence >= minConfidence ? secondaryCue : '';

    return StateOutput(
      internalState: _currentState,
      spokenText: mapData.$1,
      audioStem: mapData.$2,
      confidence: confidence,
      riskScore: riskScore,
      secondaryCue: cue,
    );
  }

  void reset() {
    _currentState = SafetyState.uncertain;
    _pendingState = SafetyState.uncertain;
    _pendingCount = 0;
  }
}
